lib/ecto/adapters/postgres.ex

```diff
@@ -145,8 +145,67 @@ defmodule Ecto.Adapters.Postgres do
     end
   end
 
+  defp run_with_psql(database, sql_command) do
+    args = ["--quiet",
+            "--set", "ON_ERROR_STOP=1",
+            "--set", "VERBOSITY=verbose",
+            "--no-psqlrc",
+            "-d", "template1",
+            "-c", sql_command]
+    run_with_cmd("psql", database, args)
+  end
+
+  defp run_with_cmd(cmd, opts, opt_args) do
+    unless System.find_executable(cmd) do
+      raise "could not find executable `#{cmd}` in path, " <>
+            "please guarantee it is available before running ecto commands"
+    end
+
+    env =
+      if password = opts[:password] do
+        [{"PGPASSWORD", password}]
+      else
+        []
+      end
+    env = [{"PGCONNECT_TIMEOUT", "10"} | env]
+
+    args = []
+
+    if username = opts[:username] do
+      args = ["-U", username|args]
+    end
+
+    if port = opts[:port] do
+      args = ["-p", to_string(port)|args]
+    end
+
+    host = opts[:hostname] || System.get_env("PGHOST") || "localhost"
+    args = ["--host", host|args]
+
+    args = args ++ opt_args
+    System.cmd(cmd, args, env: env, stderr_to_stdout: true)
+  end
+
   @doc false
   def supports_ddl_transaction? do
     true
   end
+
+  @doc false
+  def database_info do
+    { output, _ } = System.cmd("psql", ["--version"])
+    [ _, _, version ] = Regex.run(~r/^(\w+) \(\w+\) (.+)$/, output)
+    %{ type: "psql", version: version }
+  end
+
+  @doc false
+  def storage_dump(config) do
+    { output, _ } = run_with_cmd("pg_dump", config, ["-s", "-x", "-i", "-O", config[:database]])
```

Comment:
Also, what do you think about expanding these flags to their full names so it's more clear to readers what they are doing? `--schema-only`, `--no-acl`, `--no-owner`?